apps/meteor/server/modules/core-apps/mention.module.ts

```diff
@@ -0,0 +1,115 @@
+import type { IMessage } from '@rocket.chat/core-typings';
+import { Subscriptions, Messages } from '@rocket.chat/models';
+import { Meteor } from 'meteor/meteor';
+
+import { processWebhookMessage } from '../../../app/lib/server/functions/processWebhookMessage';
+import { addUsersToRoomMethod } from '../../../app/lib/server/methods/addUsersToRoom';
+import { i18n } from '../../lib/i18n';
+import { roomCoordinator } from '../../lib/rooms/roomCoordinator';
+
+const retrieveMentionsFromPayload = (stringifiedMentions: string): Exclude<IMessage['mentions'], undefined> => {
+	try {
+		const mentions = JSON.parse(stringifiedMentions);
+		if (!Array.isArray(mentions) || !mentions.length || !('username' in mentions[0])) {
+			throw new Error('Invalid payload');
+		}
+		return mentions;
+	} catch (error) {
+		throw new Error('Invalid payload');
+	}
+};
+
+export class MentionModule implements IUiKitCoreApp {
+	appId = 'mention-core';
+
+	async blockAction(payload: any): Promise<any> {
+		const {
+			actionId,
+			payload: { value: stringifiedMentions, blockId: referenceMessageId },
+		} = payload;
+
+		const mentions = retrieveMentionsFromPayload(stringifiedMentions);
+
+		const usernames = mentions.map(({ username }) => username);
+
+		const message = await Messages.findOneById(referenceMessageId);
+
+		if (!message) {
+			throw new Error('Mention bot - Failed to retrieve message information');
+		}
+
+		if (actionId === 'dismiss') {
+			void api.broadcast('notify.ephemeralMessage', payload.user._id, payload.room, {
+				msg: i18n.t(
+					'You_mentioned___mentions__but_theyre_not_in_this_room',
+					{
+						mentions: `@${usernames.join(', @')}`,
+					},
+					payload.user.language,
+				),
+				_id: payload.message,
+				tmid: message.tmid,
+				mentions,
+			});
+			return;
+		}
+
+		if (actionId === 'add-users') {
+			void addUsersToRoomMethod(payload.user._id, { rid: payload.room, users: usernames as string[] }, payload.user);
+			void api.broadcast('notify.ephemeralMessage', payload.user._id, payload.room, {
+				msg: i18n.t(
+					'You_mentioned___mentions__but_theyre_not_in_this_room',
+					{
+						mentions: `@${usernames.join(', @')}`,
+					},
+					payload.user.language,
+				),
+				tmid: message.tmid,
+				_id: payload.message,
+				mentions,
+			});
+			return;
+		}
+
+		if (actionId === 'share-message') {
+			const sub = await Subscriptions.findOneByRoomIdAndUserId(payload.room, payload.user._id);
+			// this should exist since the event is fired from withing the room (e.g the user sent a message)
+			if (!sub) {
+				throw new Error('Mention bot - Failed to retrieve room information');
+			}
+
+			const roomPath = roomCoordinator.getRouteLink(sub.t, { rid: sub.rid, name: sub.name });
+			if (!roomPath) {
+				throw new Error('Mention bot - Failed to retrieve path to room');
+			}
+
+			const link = new URL(Meteor.absoluteUrl(roomPath));
+			link.searchParams.set('msg', message._id);
+			const text = `[ ](${link.toString()})`;
```

Comment:
When mentioning someone in private rooms and selecting the share-message (via DM) option, the user who receives the notification about the mention, when clicking on this link is redirected to a `room not found` page (expected behavior, since they are not part of the private room), I just want to make sure this is the expected UX we want. 😬

Reply:
So, this was also pointed out from design recently. We introduced a message together with the quote so that it is clearer that you're not part of the room.

Being redirected to a `room not found` is now being dealt as separate bug, since sharing a message through the context menu or copying a message link and sending it in other channels produce the same behaviour.